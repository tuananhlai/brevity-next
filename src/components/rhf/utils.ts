import { ControllerProps } from "react-hook-form";

export interface RHFFieldProps
  extends Omit<ControllerProps, "disabled" | "render"> {
  // Rename `disabled` prop to `isDisabled` so that the naming convention
  // is similar between normal design system components and `Form*` components.
  isDisabled?: boolean;
}

/**
 * Omit all form-related props from design system components so that they can be replaced
 * with props retrieved from React Hook Form Controller.
 *
 * @example
 * interface FormTextFieldProps extends OmitAriaFormProps<TextFieldProps> {}
 */
export type OmitAriaFormProps<T> = Omit<
  T,
  | "name"
  | "errorMessage"
  | "isInvalid"
  | "value"
  | "defaultValue"
  | "isDisabled"
  | "selectedKey"
  | "defaultSelectedKey"
>;
