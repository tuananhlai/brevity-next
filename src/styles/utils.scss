@use "sass:string";

$breakpoint-sm: 640px;
$breakpoint-md: 768px;
$breakpoint-lg: 1024px;
$breakpoint-xl: 1280px;
$breakpoint-2xl: 1536px;

@mixin screen-sm {
  @media (min-width: $breakpoint-sm) {
    @content;
  }
}

@mixin screen-md {
  @media (min-width: $breakpoint-md) {
    @content;
  }
}

@mixin screen-lg {
  @media (min-width: $breakpoint-lg) {
    @content;
  }
}

@mixin screen-xl {
  @media (min-width: $breakpoint-xl) {
    @content;
  }
}

@mixin screen-2xl {
  @media (min-width: $breakpoint-2xl) {
    @content;
  }
}

@function alpha($color, $alphaValue) {
  @return string.unquote(
    "color-mix(in srgb, #{$color} #{$alphaValue}%, transparent)"
  );
}

@mixin dark-mode {
  &:where([data-mode="dark"], [data-mode="dark"] *) {
    @content;
  }
}

@mixin line-clamp($lines) {
  display: -webkit-box;
  line-clamp: 3;
  -webkit-line-clamp: $lines;
  -webkit-box-orient: vertical;
  overflow: hidden;
}

@mixin outline-primary {
  outline: 2px solid var(--bw-color-outline-primary);
}

@mixin text-xs {
  font-size: var(--bw-font-xs);
  line-height: 1.125rem; /* 18px */
}

@mixin text-sm {
  font-size: var(--bw-font-sm);
  line-height: 1.25rem; /* 20px */
}

@mixin text-md {
  font-size: var(--bw-font-base);
  line-height: 1.5rem; /* 24px */
}

@mixin text-lg {
  font-size: var(--bw-font-lg);
  line-height: 1.75rem; /* 28px */
}

@mixin text-xl {
  font-size: var(--bw-font-xl);
  line-height: 1.875rem; /* 30px */
}

@mixin display-xs {
  font-size: var(--bw-font-2xl);
  line-height: 2rem; /* 32px */
}

@mixin display-sm {
  font-size: var(--bw-font-3xl);
  line-height: 2.375rem; /* 38px */
}

@mixin display-md {
  font-size: var(--bw-font-4xl);
  line-height: 2.75rem; /* 44px */
}

@mixin display-lg {
  font-size: var(--bw-font-5xl);
  line-height: 3.75rem; /* 60px */
}

@mixin display-xl {
  font-size: var(--bw-font-6xl);
  line-height: 4.5rem; /* 72px */
}

@mixin display-2xl {
  font-size: var(--bw-font-7xl);
  line-height: 5.625rem; /* 90px */
}
