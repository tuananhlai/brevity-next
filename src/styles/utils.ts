/**
 * Creates a new color with the given color and alpha value.
 * @param color
 * @param alphaValue The alpha value (0 to 100)
 */
export const alpha = (color: string, alphaValue: number) => {
  return `rgb(from ${color} r g b / ${alphaValue}%)`;
};

export const darkModeSelector =
  '&:where([data-mode="dark"], [data-mode="dark"] *)';
