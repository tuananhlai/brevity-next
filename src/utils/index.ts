import { FieldErrorProps } from "react-aria-components";

export type ReplaceAriaRenderProps<T> = Omit<
  T,
  "style" | "className" | "children"
> & {
  style?: React.CSSProperties;
  className?: string;
  children?: React.ReactNode;
};

export type FieldsetProps = {
  label?: React.ReactNode;
  description?: React.ReactNode;
  errorMessage?: FieldErrorProps["children"];
};
