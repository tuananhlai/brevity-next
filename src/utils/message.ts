/** @fileoverview Commonly displayed messages within the application. */
import { msg } from "@lingui/macro";

export const somethingWentWrong = msg`Something went wrong`;
