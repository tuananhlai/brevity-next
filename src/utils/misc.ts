import { FieldErrorProps } from "react-aria-components";

/**
 * Replace React Aria Component's render props with more conventional props.
 */
export type ReplaceAriaRenderProps<T> = Omit<
  T,
  "style" | "className" | "children"
> & {
  style?: React.CSSProperties;
  className?: string;
  children?: React.ReactNode;
};

export type FieldsetProps = {
  label?: React.ReactNode;
  description?: React.ReactNode;
  errorMessage?: FieldErrorProps["children"];
};

/** The placeholder href for links that are not yet implemented. */
export const TODO_HREF = "#";

const APP_NAME = "Brevity";

/**
 * Return the page title (the text to be displayed in the browser tab) given the heading of a page.
 *
 * @example
 * ```ts
 * getPageTitle("Home"); // "Home | Brevity"
 * ```
 */
export const getPageTitle = (pageHeading: string) => {
  return `${pageHeading} | ${APP_NAME}`;
};

/**
 * Return a random avatar URL based on the given seed.
 * Same seed will always generate the same avatar.
 */
export const generateAvatarURL = (seed: string): string => {
  return `https://api.dicebear.com/9.x/glass/jpg?seed=${seed}`;
};
