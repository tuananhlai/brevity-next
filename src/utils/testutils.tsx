import { i18n } from "@lingui/core";
import { I18nProvider } from "@lingui/react";
import { render } from "@testing-library/react";
import en from "@/locales/en";

i18n.load({
  en: en,
});
i18n.activate("en");

const TestProviders = ({ children }: { children: React.ReactNode }) => {
  return <I18nProvider i18n={i18n}>{children}</I18nProvider>;
};

/**
 * Render the component with the i18n provider.
 */
const renderWithProviders = (children: React.ReactNode) => {
  return render(children, {
    wrapper: TestProviders,
  });
};

export { renderWithProviders };
