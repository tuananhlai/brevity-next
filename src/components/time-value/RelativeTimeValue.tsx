import { useLingui } from "@lingui/react";
import { getRelativeTimeString } from "@/components/time-value/utils";

export interface RelativeTimeValueProps {
  dateTime: Date;
  className?: string;
}

export const RelativeTimeValue: React.FC<RelativeTimeValueProps> = (props) => {
  const { dateTime, className } = props;
  const { i18n } = useLingui();

  return (
    <time dateTime={dateTime.toISOString()} className={className}>
      {getRelativeTimeString(dateTime, i18n.locale)}
    </time>
  );
};
