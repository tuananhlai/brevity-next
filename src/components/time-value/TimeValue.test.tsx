import { screen } from "@testing-library/react";
import { renderWithProviders } from "@/utils/testutils";
import { RelativeTimeValue } from "./RelativeTimeValue";

jest.useFakeTimers().setSystemTime(new Date("2020-01-01"));

describe("RelativeTimeValue", () => {
  it("should match snapshot", () => {
    renderWithProviders(
      <div data-testid="test-container">
        {Object.entries(timeStamps).map(([key, value]) => (
          <RelativeTimeValue key={key} dateTime={value} />
        ))}
      </div>,
    );

    expect(screen.getByTestId("test-container")).toMatchSnapshot();
  });
});

const durationInMs = {
  second: 1000,
  minute: 1000 * 60,
  hour: 1000 * 60 * 60,
  day: 1000 * 60 * 60 * 24,
  week: 1000 * 60 * 60 * 24 * 7,
  month: 1000 * 60 * 60 * 24 * 30,
  year: 1000 * 60 * 60 * 24 * 365,
};

const timeStamps = {
  now: new Date(),
  secondAgo: new Date(Date.now() - durationInMs.second),
  twoSecondsAgo: new Date(Date.now() - durationInMs.second * 2),
  minuteAgo: new Date(Date.now() - durationInMs.minute),
  twoMinutesAgo: new Date(Date.now() - durationInMs.minute * 2),
  hourAgo: new Date(Date.now() - durationInMs.hour),
  twoHoursAgo: new Date(Date.now() - durationInMs.hour * 2),
  dayAgo: new Date(Date.now() - durationInMs.day),
  twoDaysAgo: new Date(Date.now() - durationInMs.day * 2),
  weekAgo: new Date(Date.now() - durationInMs.week),
  twoWeeksAgo: new Date(Date.now() - durationInMs.week * 2),
  monthAgo: new Date(Date.now() - durationInMs.month),
  twoMonthsAgo: new Date(Date.now() - durationInMs.month * 2),
  yearAgo: new Date(Date.now() - durationInMs.year),
  twoYearsAgo: new Date(Date.now() - durationInMs.year * 2),
  secondInFuture: new Date(Date.now() + durationInMs.second),
  twoSecondsInFuture: new Date(Date.now() + durationInMs.second * 2),
  minuteInFuture: new Date(Date.now() + durationInMs.minute),
  twoMinutesInFuture: new Date(Date.now() + durationInMs.minute * 2),
  hourInFuture: new Date(Date.now() + durationInMs.hour),
  twoHoursInFuture: new Date(Date.now() + durationInMs.hour * 2),
  dayInFuture: new Date(Date.now() + durationInMs.day),
  twoDaysInFuture: new Date(Date.now() + durationInMs.day * 2),
  weekInFuture: new Date(Date.now() + durationInMs.week),
  twoWeeksInFuture: new Date(Date.now() + durationInMs.week * 2),
  monthInFuture: new Date(Date.now() + durationInMs.month),
  twoMonthsInFuture: new Date(Date.now() + durationInMs.month * 2),
  yearInFuture: new Date(Date.now() + durationInMs.year),
  twoYearsInFuture: new Date(Date.now() + durationInMs.year * 2),
};
