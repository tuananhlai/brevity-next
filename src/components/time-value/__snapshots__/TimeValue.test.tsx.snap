// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelativeTimeValue should match snapshot 1`] = `
<div
  data-testid="test-container"
>
  <time
    datetime="2020-01-01T00:00:00.000Z"
  >
    now
  </time>
  <time
    datetime="2019-12-31T23:59:59.000Z"
  >
    1 second ago
  </time>
  <time
    datetime="2019-12-31T23:59:58.000Z"
  >
    2 seconds ago
  </time>
  <time
    datetime="2019-12-31T23:59:00.000Z"
  >
    1 minute ago
  </time>
  <time
    datetime="2019-12-31T23:58:00.000Z"
  >
    2 minutes ago
  </time>
  <time
    datetime="2019-12-31T23:00:00.000Z"
  >
    1 hour ago
  </time>
  <time
    datetime="2019-12-31T22:00:00.000Z"
  >
    2 hours ago
  </time>
  <time
    datetime="2019-12-31T00:00:00.000Z"
  >
    yesterday
  </time>
  <time
    datetime="2019-12-30T00:00:00.000Z"
  >
    2 days ago
  </time>
  <time
    datetime="2019-12-25T00:00:00.000Z"
  >
    last week
  </time>
  <time
    datetime="2019-12-18T00:00:00.000Z"
  >
    2 weeks ago
  </time>
  <time
    datetime="2019-12-02T00:00:00.000Z"
  >
    last month
  </time>
  <time
    datetime="2019-11-02T00:00:00.000Z"
  >
    2 months ago
  </time>
  <time
    datetime="2019-01-01T00:00:00.000Z"
  >
    last year
  </time>
  <time
    datetime="2018-01-01T00:00:00.000Z"
  >
    2 years ago
  </time>
  <time
    datetime="2020-01-01T00:00:01.000Z"
  >
    in 1 second
  </time>
  <time
    datetime="2020-01-01T00:00:02.000Z"
  >
    in 2 seconds
  </time>
  <time
    datetime="2020-01-01T00:01:00.000Z"
  >
    in 1 minute
  </time>
  <time
    datetime="2020-01-01T00:02:00.000Z"
  >
    in 2 minutes
  </time>
  <time
    datetime="2020-01-01T01:00:00.000Z"
  >
    in 1 hour
  </time>
  <time
    datetime="2020-01-01T02:00:00.000Z"
  >
    in 2 hours
  </time>
  <time
    datetime="2020-01-02T00:00:00.000Z"
  >
    tomorrow
  </time>
  <time
    datetime="2020-01-03T00:00:00.000Z"
  >
    in 2 days
  </time>
  <time
    datetime="2020-01-08T00:00:00.000Z"
  >
    next week
  </time>
  <time
    datetime="2020-01-15T00:00:00.000Z"
  >
    in 2 weeks
  </time>
  <time
    datetime="2020-01-31T00:00:00.000Z"
  >
    next month
  </time>
  <time
    datetime="2020-03-01T00:00:00.000Z"
  >
    in 2 months
  </time>
  <time
    datetime="2020-12-31T00:00:00.000Z"
  >
    next year
  </time>
  <time
    datetime="2021-12-31T00:00:00.000Z"
  >
    in 2 years
  </time>
</div>
`;
